qiskit/transpiler/passes/synthesis/unitary_synthesis.py

```diff
@@ -82,31 +103,90 @@ def run(self, dag: DAGCircuit) -> DAGCircuit:
             elif len(node.qargs) == 2:
                 if decomposer2q is None:
                     continue
-                synth_dag = circuit_to_dag(
-                    decomposer2q(node.op.to_matrix(), basis_fidelity=self._approximation_degree)
-                )
+                synth_dag, wires = self._synth_natural_direction(node, wires, dag, decomposer2q)
             else:
                 synth_dag = circuit_to_dag(isometry.Isometry(node.op.to_matrix(), 0, 0).definition)
 
-            dag.substitute_node_with_dag(node, synth_dag)
+            dag.substitute_node_with_dag(node, synth_dag, wires=wires)
 
         return dag
+
+    def _synth_natural_direction(self, node, wires, dag, decomposer2q):
+        layout = self.property_set["layout"]
+        natural_direction = None
+        synth_direction = None
+        physical_gate_fidelity = None
+        if self._natural_direction and layout and self._coupling_map:
+            neighbors = self._coupling_map.neighbors(dag.qubits.index(node.qargs[0]))
+            zero_one = dag.qubits.index(node.qargs[1]) in neighbors
+            neighbors = self._coupling_map.neighbors(dag.qubits.index(node.qargs[1]))
+            one_zero = dag.qubits.index(node.qargs[0]) in neighbors
+            if zero_one and not one_zero:
+                natural_direction = [0, 1]
+            if one_zero and not zero_one:
+                natural_direction = [1, 0]
+
+        elif self._natural_direction and layout and self._backend_props:
+            len_0_1 = inf
+            len_1_0 = inf
+            try:
+                len_0_1 = self._backend_props.gate_length(
+                    "cx", [dag.qubits.index(node.qargs[0]), dag.qubits.index(node.qargs[1])]
+                )
+            except BackendPropertyError:
+                pass
+            try:
+                len_1_0 = self._backend_props.gate_length(
+                    "cx", [dag.qubits.index(node.qargs[1]), dag.qubits.index(node.qargs[0])]
+                )
+            except BackendPropertyError:
+                pass
+
+            if len_0_1 < len_1_0:
+                natural_direction = [0, 1]
+            elif len_1_0 < len_0_1:
+                natural_direction = [1, 0]
+            if natural_direction:
+                physical_gate_fidelity = 1 - self._backend_props.gate_error(
+                    "cx", [dag.qubits.index(node.qargs[i]) for i in natural_direction]
+                )
+        basis_fidelity = self._approximation_degree or physical_gate_fidelity
+        su4_mat = node.op.to_matrix()
+        synth_circ = decomposer2q(su4_mat, basis_fidelity=basis_fidelity)
+        synth_dag = circuit_to_dag(synth_circ)
+
+        # if a natural direction exists but the synthesis is in the opposite direction,
+        # resynthesize a new operator which is the original conjugated by swaps.
+        # this new operator is doubly mirrored from the original and is locally equivalent.
+        if synth_dag.two_qubit_ops():
```

Comment:
I recently learned that whole `if`s like this can be rewritten as

```
synth_direction = next([synth_dag.qubits.index(q) for q in op.qargs] for op in synth_dag.two_qubit_ops(), None)
```

It's not much more readable in this case, given the inner map. Still, FYI.

Reply:
thanks, although it might take some getting used to for me for it to appear more readable. Also, in this case I don't actually do a double loop but rather assume all two qubit ops have the same direction as the first. I'll have to check whether this is a valid assumption...I think this is the way TwoQubitBasisDecomposer works but not sure if it promises that.

Reply:
I don't think it's explicitly promised, but it does seem to be implicitly assumed, and not just by you. I wouldn't worry about that.

There's no double loop here either, since the outer iter only gets `next`ed once, then discarded. Again, just food for thought; far from a critical suggestion.